Mathlib/NumberTheory/EllipticDivisibilitySequence.lean

```diff
@@ -65,26 +75,506 @@ elliptic, divisibility, sequence
+  rw [Perm.coe_mul, ← Function.comp.assoc, hτ, hσ, map_mul, mul_comm, mul_smul]
+
+private lemma rel₄Fin4_perm' (σ : Perm (Fin 4)) (t) :
+    Perm.sign σ • rel₄Fin4 W (t ∘ σ) = rel₄Fin4 W t := by
+  rw [rel₄Fin4_perm neg, ← mul_smul, Int.units_mul_self, one_smul]
+
+variable (zero : W 0 = 0)
+
+/-! `rel₄` is trivial when two indices are equal. -/
+
+private lemma rel₄_same₀₁ (m r s : ℤ) : rel₄ W m m r s = 0 := by
+  simp_rw [rel₄, addMulSub_self W zero]; ring
+
+private lemma rel₄_same₁₂ (m n s : ℤ) : rel₄ W m n n s = 0 := by
+  simp_rw [rel₄, addMulSub_self W zero]; ring
+
+private lemma rel₄_same₂₃ (m n r : ℤ) : rel₄ W m n r r = 0 := by
+  simp_rw [rel₄, addMulSub_self W zero]; ring
+
+variable (one : W 1 ∈ R⁰) (two : W 2 ∈ R⁰)
+  (oddRec : ∀ m ≥ 2, OddRec W m) (evenRec : ∀ m ≥ 3, EvenRec W m)
+
+/-- The four-index `rel₄` relations follow from
+the single-index `oddRec` and `evenRec` recursive relations. -/
+private theorem rel₄_of_oddRec_evenRec {a b c d : ℤ} (same : HaveSameParity₄ a b c d) :
+    rel₄ W a b c d = 0 := by
+  let t := ![|a|, |b|, |c|, |d|]
+  have nonneg i : 0 ≤ t i := by fin_cases i <;> exact abs_nonneg _
+  let σ := Fin.revPerm.trans (Tuple.sort t)
+  have anti : Antitone (t ∘ σ) := by
+    simp_rw [σ, coe_trans, ← Function.comp.assoc]
+    exact (Tuple.monotone_sort t).comp_antitone fun _ _ ↦ Fin.rev_le_rev.mpr
+  clear_value σ -- otherwise, unifying `t (σ i)` with `(t ∘ σ) i` is extremely slow
+  rw [← rel₄_abs neg]; change rel₄Fin4 W t = 0
+  rw [← rel₄Fin4_perm' neg σ, rel₄Fin4]; simp_rw [Function.comp]
+  by_cases h₃₂ : t (σ 3) = t (σ 2); · rw [h₃₂, rel₄_same₂₃ zero, smul_zero]
+  by_cases h₂₁ : t (σ 2) = t (σ 1); · rw [h₂₁, rel₄_same₁₂ zero, smul_zero]
+  by_cases h₁₀ : t (σ 1) = t (σ 0); · rw [h₁₀, rel₄_same₀₁ zero, smul_zero]
+  rw [rel₄_of_anti_oddRec_evenRec one two oddRec evenRec (same.abs.perm _ _), smul_zero]
+  exact ⟨nonneg _, (anti <| by decide).lt_of_ne h₃₂,
+    (anti <| by decide).lt_of_ne h₂₁, (anti <| by decide).lt_of_ne h₁₀⟩
+
+/-- An ℕ-indexed sequence defined recursively by the even-odd recurrence, after extension to
+all integers by symmetry (to make an odd function), is an elliptic sequence, provided its
+first two terms are not zero divisors. -/
+theorem _root_.IsEllSequence.of_oddRec_evenRec : IsEllSequence W := fun m n r ↦ by
+  rw [rel₃_iff_rel₄_eq_zero, rel₄_of_oddRec_evenRec neg zero one two oddRec evenRec]
+  refine ⟨?_, ?_, ?_⟩ <;> simp only [negOnePow_two_mul, negOnePow_zero]
+
+end Perm
+
+end EllSequence
+
+open EllSequence
 
 /-- The proposition that a sequence indexed by integers is a divisibility sequence. -/
 def IsDivSequence : Prop :=
-  ∀ m n : ℕ, m ∣ n → W m ∣ W n
+  ∀ m n : ℤ, m ∣ n → W m ∣ W n
 
 /-- The proposition that a sequence indexed by integers is an EDS. -/
 def IsEllDivSequence : Prop :=
   IsEllSequence W ∧ IsDivSequence W
 
 lemma isEllSequence_id : IsEllSequence id :=
-  fun _ _ _ => by simp only [id_eq]; ring1
+  fun _ _ _ ↦ by simp only [Rel₃, id_eq]; ring1
```

Comment:
What is the recommended symbol to use here? I always used `=>` but you always use `↦`.